context/wallet.tsx

```diff
@@ -0,0 +1,151 @@
+import { AccountInfo, LocalStorage } from "@airgap/beacon-sdk";
+import { BeaconWallet } from "@taquito/beacon-wallet";
+import { TezosToolkit } from "@taquito/taquito";
+import React, { createContext, useEffect, useReducer } from "react";
+import { PREFERED_NETWORK } from "../context/config";
+import { RPC_URL } from "./config";
+import { BeaconSigner } from "./signer";
+import { useTezosToolkit } from "./tezos-toolkit";
+
+type WalletState = {
+  wallet: BeaconWallet | undefined;
+  userAccount: AccountInfo | undefined;
+  userAddress: string | undefined;
+  userBalance: number | undefined;
+};
+
+type WalletContextType = {
+  connectWallet: () => void;
+  disconnectWallet: () => void;
+  state: WalletState;
+};
+
+const createWalletConnection = (tezos: TezosToolkit) => {
+  const wallet = new BeaconWallet({
+    name: "TzSafe",
+    preferredNetwork: PREFERED_NETWORK,
+    storage: new LocalStorage("WALLET"),
+  });
+
+  tezos.setProvider({
+    rpc: RPC_URL,
+    wallet,
+  });
+  tezos.setSignerProvider(new BeaconSigner(wallet));
+
+  return wallet;
+};
+
+const initialState: WalletState = {
+  wallet: undefined,
+  userAccount: undefined,
+  userAddress: undefined,
+  userBalance: undefined,
+};
+
+export const WalletContext = createContext<WalletContextType>({
+  state: initialState,
```

Comment:
why wrapping the initialState in an object ? 🤔 

Not wrapping it would allow to directl access the userAddress like so:

```js
const { userAddress } = useWallet();
```

wdyt?

Reply:
Yes we can do this. I wrote this because I wanted to split " state logic " to " update functions ".
But it can be merge :)